syntax = "proto3";

package federation;

// 模型参数消息
message ModelParameters {
    // 参数名称到参数值的映射
    map<string, bytes> parameters = 1;
}

// 训练指标消息
message TrainingMetrics {
    float accuracy = 1;
    float precision = 2;
    float recall = 3;
    float f1 = 4;
    float auc_roc = 5;
}

// 客户端信息消息
message ClientInfo {
    string client_id = 1;
    string model_type = 2;
    int32 data_size = 3;
}

// 服务定义
service FederatedLearning {
    // 客户端注册
    rpc RegisterClient(ClientInfo) returns (RegisterResponse) {}
    
    // 提交本地参数
    rpc SubmitParameters(ParameterSubmission) returns (ParameterResponse) {}
    
    // 获取全局参数
    rpc GetGlobalParameters(ParameterRequest) returns (ModelParameters) {}
    
    // 提交训练指标
    rpc SubmitMetrics(MetricsSubmission) returns (MetricsResponse) {}
}

// 注册响应
message RegisterResponse {
    bool success = 1;
    string message = 2;
}

// 参数提交
message ParameterSubmission {
    string client_id = 1;
    ModelParameters parameters = 2;
    TrainingMetrics metrics = 3;
}

// 参数响应
message ParameterResponse {
    bool success = 1;
    string message = 2;
}

// 参数请求
message ParameterRequest {
    string client_id = 1;
}

// 指标提交
message MetricsSubmission {
    string client_id = 1;
    TrainingMetrics metrics = 2;
}

// 指标响应
message MetricsResponse {
    bool success = 1;
    string message = 2;
} 